consul/src/main/java/com/linecorp/armeria/client/consul/ConsulEndpointGroup.java

```diff
@@ -0,0 +1,117 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.ClientRequestContextCaptor;
+import com.linecorp.armeria.client.Clients;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.DynamicEndpointGroup;
+import com.linecorp.armeria.client.endpoint.EndpointGroup;
+import com.linecorp.armeria.internal.consul.ConsulClient;
+
+import io.netty.channel.EventLoop;
+
+/**
+ * A Consul-based {@link EndpointGroup} implementation that retrieves the list of {@link Endpoint}s
+ * from Consul using <a href="https://www.consul.io/api">Consul's RESTful HTTP API</a> and
+ * updates the {@link Endpoint}s periodically.
+ */
+public final class ConsulEndpointGroup extends DynamicEndpointGroup {
+
+    private static final Logger logger = LoggerFactory.getLogger(ConsulEndpointGroup.class);
+
+    /**
+     * Returns a {@link ConsulEndpointGroup} with the specified {@code serviceName}.
+     * The returned {@link ConsulEndpointGroup} will retrieve the list of {@link Endpoint}s from
+     * a local Consul agent(using default Consul service port).
+     */
+    public static ConsulEndpointGroup of(String serviceName) {
+        return builder(serviceName).build();
+    }
+
+    /**
+     * Returns a newly-created {@link ConsulEndpointGroupBuilder} with the specified {@code serviceName}.
+     */
+    public static ConsulEndpointGroupBuilder builder(String serviceName) {
+        return new ConsulEndpointGroupBuilder(serviceName);
+    }
+
+    private final ConsulClient consulClient;
+    private final String serviceName;
+    private final long registryFetchIntervalSeconds;
+    private final boolean useHealthyEndpoints;
+
+    @Nullable
+    private volatile ScheduledFuture<?> scheduledFuture;
+
+    ConsulEndpointGroup(ConsulClient consulClient, String serviceName, long registryFetchIntervalSeconds,
+                        boolean useHealthyEndpoints) {
+        this.consulClient = consulClient;
+        this.serviceName = serviceName;
+        this.registryFetchIntervalSeconds = registryFetchIntervalSeconds;
+        this.useHealthyEndpoints = useHealthyEndpoints;
+
+        update();
+    }
+
+    private void update() {
+        if (isClosing()) {
+            return;
+        }
+
+        final CompletableFuture<List<Endpoint>> response;
+        final EventLoop eventLoop;
+        try (ClientRequestContextCaptor captor = Clients.newContextCaptor()) {
+            if (useHealthyEndpoints) {
+                response = consulClient.healthyEndpoints(serviceName);
+            } else {
+                response = consulClient.endpoints(serviceName);
+            }
+            eventLoop = captor.get().eventLoop().withoutContext();
+        }
+
+        response.handle((endpoints, cause) -> {
+            if (cause != null) {
```

Comment:
How about checking `isClosing()` here?

Reply:
You mean it needs to check again after query?
If so, is this good?
```
        }

        if (isClosing()) {
            return;
        }

        response.handle((endpoints, cause) -> {
            if (cause != null) {
```

Reply:
What I meant is:
```java
response.handle((endpoints, cause) -> {
    if (isClosing()) {
        return;
    }
    if (cause != null) {
    ...
```

Reply:
Oop! I forgot that response is a future.
Ok.